### PR TITLE
ENH: Add recursive least squares model

### LINKED ISSUE #2759: ENH: Recursive least squares (via statespace)
This comes up enough and would be easy enough that it may make sense to add a new recursive least squares (RLS) model / class. This issue is just to poll for whether or not it should be included and what features people would like that I haven't thought of.

Technically, this class would simply be a wrapper around the `SARIMAX` class which can compute RLS estimates already (optionally with SARIMA errors). So if we do create a new RLS class, it would be more for "marketing" purposes and for RLS-specific output than for functionality.

### PR DESCRIPTION
**Pull request by ChadFulton**
Closes #2759.

This adds a recursive least squares model at `statsmodels.regression.rls`, where computations are handled by the Kalman filter. In addition to the recursive coefficient estimates, it includes CUSUM and CUSUM of squares statistics and diagnostic plots.

See the example notebook rendered here: http://nbviewer.jupyter.org/gist/ChadFulton/7a63fbfca03276a04f2a

### CODE REVIEW

**Comment by josef-pkt** [comment]
very good, cusum test is somewhere in diagnostic `breaks_xxx` (confidence bounds compared to GRETL, IIRC)

try to find longer names  `rls` is robust least squares, restricted least squares, random least squares, ...

(IIRC, we have RLS for restricted least squares in a PR, but I'm using `constrained` now instead of restricted as general naming convention.)

**Comment by ChadFulton** [comment]
How about `statsmodels.regression.recursive_ls` and class `RecursiveLS`?

I have unit tests against the existing CUSUM `breaks_` functions.

**Comment by josef-pkt** [comment]
names with `recursive` sound good.

(I might not be able to review very soon because I got stuck in sandwiches. sticky exposure :)

**Comment by ChadFulton** [comment]
No problem on whenever it's convenient to review; thanks.

**Comment by ChadFulton** [comment]
One thing I wanted to mention is that presumably this would be a model where allowing the formula interface would make sense, but I don't know much about getting that working.

**Comment by josef-pkt** [comment]
just try out `from_formula` in the same way as in OLS.
Because your class has a standard `__init__` signature, everything might/should just work automatically by inheritance.

`RecursiveLS.from_formula(formula, data)....`

File: statsmodels/regression/recursive_ls.py [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+        )
+        if self.smoothed_state is not None:
+            out.smoothed = self.smoothed_state[start:end]
+        if self.smoothed_state_cov is not None:
+            out.smoothed_cov = (
+                self.smoothed_state_cov[start:end, start:end])
+        return out
+
+    @cache_readonly
+    def recursive_resid(self):
```
Comment (josef-pkt) [inline statsmodels/regression/recursive_ls.py]: postfix naming
`resid_recursive`

File: statsmodels/regression/recursive_ls.py [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+        # See Harvey (1989) section 5.4; he defines the standardized
+        # innovations in 5.4.1, but they have non-unit variance, whereas
+        # the standardized forecast errors assume unit variance. To convert
+        # to Harvey's definition, we need to multiply by the standard
+        # deviation.
+        return (self.filter_results.standardized_forecasts_error.squeeze() *
+                self.filter_results.obs_cov[0, 0]**0.5)
+
+    @cache_readonly
+    def cusum(self):
```
Comment (josef-pkt) [inline statsmodels/regression/recursive_ls.py]: maybe `resid_cusum` ?

File: statsmodels/regression/recursive_ls.py [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+           Journal of the Royal Statistical Society.
+           Series B (Methodological) 37 (2): 149-92.
+
+        """
+        llb = self.loglikelihood_burn
+        return (np.cumsum(self.recursive_resid[self.loglikelihood_burn:]) /
+                np.std(self.recursive_resid[llb:], ddof=1))
+
+    @cache_readonly
+    def cusum_squares(self):
```
Comment (josef-pkt) [inline statsmodels/regression/recursive_ls.py]: I don't know a good name that would also include resid or squared resid. Maybe fine the way it is.

File: statsmodels/regression/recursive_ls.py (hunk 2 above)
Comment (josef-pkt) [inline statsmodels/regression/recursive_ls.py]: I think `cusum` is fine. Currently, `resid_xxx` always refer to a single observation.

**Comment by josef-pkt** [comment]
quick browse again.

The only change I see is to change the name of `resid_recursive`.

Inference on parameters, standard errors, wald tests, ... are currently not available, AFAICS, are they?
If not, then a comment or note in the docstring would be useful.

@ChadFulton  Given that we haven't received any feedback by potential users, I would merge this as is. We expand it in future when we get feedback or find our own additional uses.
The way it is right now it's useful also for stability checks.

**Comment by ChadFulton** [comment]
> Inference on parameters, standard errors, wald tests, ... are currently not available, AFAICS, are they?
If not, then a comment or note in the docstring would be useful.

Inference on the final recursive parameters is available as usual (the Results object is populated with the parameters and standard errors from the final estimates at the end of the sample); unless I've misunderstood your point?

**Comment by ChadFulton** [comment]
Assuming all comes back green I will aim to merge this evening, unless something else comes up.

**Comment by josef-pkt** [comment]
> Inference on the final recursive parameters is available

good, I was thinking of multivariate estimation problems where we would want to have inference for all estimated parameters, not just the final. (I don't know if there is a usecase for those.)

**Comment by ChadFulton** [comment]
If Travis and Coveralls come back green, I will merge.

**Comment by ChadFulton** [comment]
Reminder on the BDS test reminded me that before I merge I need to add to release notes.

**Comment by josef-pkt** [comment]
@ChadFulton  You could add to the release notes after merging this

I just opened the PR with my branch that creates the release file
https://github.com/statsmodels/statsmodels/pull/3005/files
I will look at it and merge it tomorrow.

there are also statespace models that need some more description
I was trying to track the status of 0.8 in #2176

**Comment by ChadFulton** [comment]
Okay, I'll add to the release notes for 0.8 when they're merged.

**Pull request merged**